x-pack/apm-server/sampling/eventstorage/storage.go

```diff
@@ -75,7 +81,7 @@ func (s *Storage) NewReadWriter() *ReadWriter {
 	s.pendingSize.Add(baseTransactionSize)
 	return &ReadWriter{
 		s:           s,
-		txn:         s.db.NewTransaction(true),
+		txn:         nil, // lazy init to avoid deadlock in storage manager
```

Comment:
[to reviewer] After fully protecting db access, there was a deadlock due to NewTransaction, changed to lazy init in ReadWriter. The alternative is to create NewReadWriter with a provided txn, but that requires more code change, and lazy init is actually more resource efficient.